src/App.ts

```diff
@@ -541,11 +544,44 @@ export default class App {
     ] as Middleware<AnyMiddlewareArgs>[]);
   }
 
-  // TODO: just make a type alias for Middleware<SlackEventMiddlewareArgs<'message'>>
-  // TODO: maybe remove the first two overloads
-  public message(...listeners: Middleware<SlackEventMiddlewareArgs<'message'>>[]): void;
-  public message(pattern: string | RegExp, ...listeners: Middleware<SlackEventMiddlewareArgs<'message'>>[]): void;
-  public message(...patternsOrMiddleware: (string | RegExp | Middleware<SlackEventMiddlewareArgs<'message'>>)[]): void {
+  /**
+   *
+   * @param listeners Middlewares that process and react to a message event
+   */
+  public message(...listeners: MessageEventMiddleware[]): void;
+  /**
+   *
+   * @param pattern Used for filtering out messages that don't match.
+   * Strings match via {@link String.prototype.includes}.
+   * @param listeners Middlewares that process and react to the message events that matched the provided patterns.
+   */
+  public message(pattern: string | RegExp, ...listeners: MessageEventMiddleware[]): void;
+  /**
+   *
+   * @param filter Middleware that can filter out messages. Generally this is done by returning before
+   * calling {@link AllMiddlewareArgs.next} if there is no match. See {@link directMention} for an example.
+   * @param pattern Used for filtering out messages that don't match the pattern. Strings match
+   * via {@link String.prototype.includes}.
+   * @param listeners Middlewares that process and react to the message events that matched the provided patterns.
+   */
+  public message(
+    filter: MessageEventMiddleware, pattern: string | RegExp, ...listeners: MessageEventMiddleware[]
+  ): void;
+  /**
+   *
+   * @param filter Middleware that can filter out messages. Generally this is done by returning before calling
+   * {@link AllMiddlewareArgs.next} if there is no match. See {@link directMention} for an example.
+   * @param listeners Middlewares that process and react to the message events that matched the provided patterns.
+   */
+  public message(filter: MessageEventMiddleware, ...listeners: MessageEventMiddleware[]): void;
+  /**
+   * This allows for further control of the filtering and response logic. Patterns and middlewares are processed in
+   * the order provided. If any patterns do not match, or a middleware does not call {@link AllMiddlewareArgs.next},
+   * all remaining patterns and middlewares will be skipped.
+   * @param patternsOrMiddleware A mix of patterns and/or middlewares.
+   */
+  public message(...patternsOrMiddleware: (string | RegExp | MessageEventMiddleware)[]): void;
```

Comment:
Curious why we need both the definition and the implementation for this one. Previously it was just the implementation.

Reply:
The typescript compiler doesn't create the type definition for the actual implementation, only the overrides. This can be seen in the current App.d.ts in the release(Below is from v3.7)

![image](https://user-images.githubusercontent.com/2661819/139513954-5275b716-adf7-4ef8-8437-dbe49bb168d0.png)

Reply:
Thanks for the explanation!